// bobflux/examples/src/mainPage.js
"use strict";
var b = require("bobril");
var m = require("bobril-m");
var g = require("bobril-g11n");
var fg = require("bobril-flexbox-grid");
var f = require("./flux");
var a = require("./mainPageActions");
var c = require("./states.cursors");
var ui = require("./userInfoComponent");
var tdp = require("./todos/todosPage");
var wnp = require("./whatNext/whatNextPage");
var ua = require("./userAccount/userAccountPage");
exports.createMainPage = f.createRouteComponent({
    init: function (ctx) {
        ctx.userInfoFactory = ui.createUserInfo(c.userInfoCursor);
    },
    render: function (ctx, me) {
        me.children = [
            m.Paper({
                children: [
                    fg.Row({
                        children: [
                            fg.Col({
                                md: 3,
                                children: m.Button({
                                    type: 1 /* Raised */,
                                    feature: b.isActive("todos") || b.isActive("default") ? 2 /* Secondary */ : 1 /* Primary */,
                                    children: g.t("Todos"),
                                    action: function () { a.goTo("todos"); }
                                })
                            }),
                            fg.Col({
                                md: 3,
                                children: m.Button({
                                    type: 1 /* Raised */,
                                    feature: b.isActive("whatNext") ? 2 /* Secondary */ : 1 /* Primary */,
                                    children: g.t("What next?"),
                                    action: function () { return a.goTo("whatNext"); }
                                })
                            }),
                            fg.Col({
                                md: 3,
                                children: m.Button({
                                    type: 1 /* Raised */,
                                    feature: b.isActive("userAccount") ? 2 /* Secondary */ : 1 /* Primary */,
                                    children: g.t("Account"),
                                    action: function () { return a.goTo("userAccount"); }
                                })
                            }),
                            fg.Col({
                                md: 1,
                                children: []
                            }),
                            fg.Col({
                                md: 2,
                                children: ctx.userInfoFactory()
                            })
                        ]
                    }),
                ]
            }),
            m.Paper({
                children: fg.Row({
                    center: 1 /* xs */,
                    children: fg.Col({
                        md: 10,
                        children: me.data.activeRouteHandler()
                    })
                })
            })
        ];
    }
});
function createRoutes() {
    return [
        b.route({
            url: "/todos",
            name: "todos",
            handler: tdp.createTodosPage(c.todosCursor)
        }),
        b.route({
            url: "/whatNext",
            name: "whatNext",
            handler: wnp.createWhatNextPage(c.whatNextCursor)
        }),
        b.route({
            url: "/userAccount",
            name: "userAccount",
            handler: ua.createUserAccountPage((_a = {},
                _a[f.defaultStateName] = c.userAccountCursor,
                _a.userInfo = c.userInfoCursor,
                _a
            ))
        }),
        b.routeDefault({
            name: "default",
            handler: tdp.createTodosPage(c.todosCursor)
        })
    ];
    var _a;
}
exports.createRoutes = createRoutes;
