// fun-model/src/actionFactory.js
"use strict";
var s = require('./store');
var d = require('./debug');
var h = require('./helpers');
var stateChanged = null;
var exceptionHandling;
exports.bootstrap = function (onStateChanged, withExceptionHandling) {
    if (withExceptionHandling === void 0) { withExceptionHandling = false; }
    stateChanged = onStateChanged;
    queueOfHandlers = [];
    exceptionHandling = withExceptionHandling;
    d.log('Action factory has been initialized.');
};
function defaultHandler(_oldValue, newValue) { return newValue; }
exports.createAction = function (cursor, handler) {
    if (handler === void 0) { handler = defaultHandler; }
    return (function (params) {
        if (stateChanged === null)
            throw 'Render callback must be set before first usage through bootstrap(defaultState, () => { yourRenderCallback(); }).';
        if (changeStateWithQueue(unifyCursor(cursor, params), handler, params)) {
            stateChanged();
            d.log('Rendering invoked...');
        }
    });
};
function unifyCursor(cursor, params) {
    return cursor.create instanceof Function ? cursor.create(params) : cursor;
}
exports.createActions = function () {
    var pairs = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        pairs[_i - 0] = arguments[_i];
    }
    return (function (params) {
        if (stateChanged === null)
            throw 'Render callback must be set before first usage through bootstrap(defaultState, () => { yourRenderCallback(); }).';
        var changed = false;
        for (var i in pairs)
            if (pairs.hasOwnProperty(i)) {
                var pair = pairs[i];
                if (changeStateWithQueue(pair.cursor, pair.handler, params))
                    changed = true;
            }
        changed && stateChanged();
    });
};
var queueOfHandlers = [];
function changeStateWithQueue(cursor, handler, params) {
    queueOfHandlers.push({ cursor: cursor, handler: handler, params: params });
    if (queueOfHandlers.length > 1)
        return false;
    var isStateChanged = false;
    while (queueOfHandlers.length > 0) {
        var n = queueOfHandlers[0];
        if (h.isFunction(exceptionHandling) ? exceptionHandling() : exceptionHandling)
            try {
                isStateChanged = changeState(n.cursor, n.handler, n.params) || isStateChanged;
            }
            catch (error) {
                d.log('Error in action handling: ', error);
            }
        else
            isStateChanged = changeState(n.cursor, n.handler, n.params) || isStateChanged;
        queueOfHandlers.shift();
    }
    isStateChanged && d.log('Global state has been changed.');
    return isStateChanged;
}
function changeState(cursor, handler, params) {
    var oldState = s.getState(cursor);
    var newState = handler(oldState, params);
    if (oldState === newState)
        return false;
    s.setState(cursor, newState);
    return true;
}
